docs/development/core/development-functional-tests.asciidoc

```diff
@@ -226,14 +226,103 @@ The `FunctionalTestRunner` comes with three built-in services:
+*** args
+**** `keys`
+***** string : A string with the keys that should be want pressed.
+***** string[] : An array of the strings that should be pressed. 
+** `browser.clickMouseButton(...args)` - Performs a left click at a given origin being either the center of an element, the viewport or the mouse's current position. 
+*** args
+**** `element`
+***** { WebElementWrapper } 
+**** `xOffset`
+***** number : number of pixels from left of origin to target click
+**** `yOffset`
+***** number : number of pixels from top of origin to target click
+** `browser.getPageSource()` - Gets the HTML loaded in the focused window/frame.
+*** returns
+**** string : HTML source
+** `browser.getLogsFor(...args)` - Get's all logs from the remote environment of the given type. 
+*** args
+**** `type`
+***** { logging.Type } - Type Object from Selenium Library to indicate what type of logs to return.
+*** returns
+**** { LogEntry[] } : Array of LogEntry items; Logs are cleared after return. 
+** `browser.takeScreenshot()` - Gets a screenshot of the focused window/frame and returns it as a base-64 encoded PNG.
+*** returns
+**** Buffer : The PNG
+** `browser.doubleClick(element)` - Performs a double left-click with the mouse at a given element. If no element specified, mouse will double-click at its current location.
+**** `element`
+***** { WebElementWrapper } 
+** `browser.switchWindowTo(...args)` - Changes the focus to a new window of the specified handle. All actions will be performed on this window going forward.
+*** args 
+**** `handle`
+***** string : Identifier for the window
+** `browser.getAllWindowHandles()` - Get the handles of all currently open windows.
+*** returns
+**** string[] : Array of all open window handles.
+** `browser.closeCurrentWindow()` - Closes the currently open window.
+** `browser.execute(fn, ...args)` - Executes the given JavaScript function within the focused window/frame.
+*** `function`
+**** string : String representation of desired function to execute.
+**** function : JavaScript function to execute.
+*** `args`
+**** any[] - An array of any type containing arguments to pass into given function.
+** `browser.setLocalStorageItem(key, value)` - Sets a value in local storage for the focused window/frame.
+*** `key`
+**** string : Key to set in local storage.
+*** `value`
+**** string : Value for given key to set in local storage.
+
 **retry:**:::
 * Source: {blob}test/functional/services/retry.js[test/functional/services/retry.js]
-* Helpers for retrying operations
-* Popular methods:
+* **A Last Resort For Retrying Operations**
+* Up until recently, `retry` was our go-to operation for waiting for a condition to be true before continuing.
+* This has been replaced by the waiting mechanism within the `find` service. Within this waiting mechanism, retries are 
+* already included. Calling `retry` directly should be avoided unless it's the only option.
+* Methods
```

Comment:
Isn't there a set of methods `waitFor`?  I think that's what we wanted to move people towards.  The advantage if I recall was that there was a string parameter where you describe what the code was waiting for to so that the debug logging was more meaningful.

Reply:
Also, regardless of the retry mechanism, I'm wondering if this is the place to describe the desired techniques for waiting for completion of one action before trying (and retrying) the next action.

Maybe some examples would be appropriate?  

Any service method or page object method that does a `click()` is potentially triggering actions that will be reflected by changes in the DOM.  In some cases the action triggers a request to Elasticsearch and the response will cause a change in the DOM.  The method that does the click is responsible for waiting for the completion of the action so that the next step of the test will work as expected.  In many cases there are loading indicators that appear, disappear, or change state to indicate the action is completed.
Cases where we don't currently have a good indication of when an action has completed are typically where we've resorted to using retry methods.
One indication that a previous action hasn't completed before the next test step is attempted is a `stale element` error.  This generally means that the test found an element and tried to click it or read some property of it (like visible text), but the element changed in between the time it was found and the time it was accessed.
Other cases can appear to just fail to do the expected action.  For example, clicking Save on a search (with no errors that it couldn't find it or click it) but the save dialog doesn't appear.